upd: Numbers page (#5503)
JetBrains/kotlin-web-site: docs/topics/numbers.md

[//]: # (title: Numbers)
[//]: # (description: Learn how to use numbers in Kotlin, including numeric types, literals, conversions, arithmetic operations, overflow, and JVM-specific behavior.)

> 

Reviewer: Extra empty line :)

The Kotlin number types represent:
* Integer values ([Byte](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/-byte/),
  [Short](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/-short/),
  [Int](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/-int/),
  and [Long](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/-long/))
* Floating-point values ([Float](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/-float/)
  and [Double](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/-double/))

> Use numbers to store and process numeric data, for example, in arithmetic, counters, measurements,

Reviewer: Probably "number types", not "numbers"? I would be amused if someone started to explain to me what numbers are for :)

and other calculations.

> ## Choose a number type

Reviewer: The paragraph above seems to me to go well with the list below, so I'm not sure a separate section is the best approach here. Also it can probably be cut down to something like:

> Use number types to store <...> other calculations. For example:
> * Use `Int` for whole numbers
> * ...

Author: I want to keep an intro as short and crisp as possible :) Plus I believe it is better for the SEO

In most cases, you can refer to the following rules to determine the
correct number type for your task:

* Use `Int` for whole numbers.
* Use `Long` for whole numbers outside the `Int` range.
* Use `Double` for decimal numbers.
* Use `Float` when lower precision is acceptable or required.
* Use `Byte` and `Short` when an API or data format requires them.

> > Kotlin also provides [](unsigned-integer-types.md).

Reviewer: Since we just explained what `Int` is for, probably worth it to say a couple of words how unsigned `Int` is different, if only for consistency.

Author: Wouldn't it overwhelm a reader? People who need such a page probably will need some time before needing such types. And for people experienced enough to need it, this note should be okay.

Reviewer: As far as I remember, unsigned ints are separate primarily because they are in Beta still, not because there's something overwhelming about them. They are unsigned, big deal :)

The previous version was also vague, maybe we should address this head on: "Kotlin also provides unsigned integer types, but they are still considered a Beta feature." — that definitely would deter those, who shouldn't waste time on digging into this.

Author: Good point, you are right! Updated the tip as suggested.

```kotlin
fun main() {
//sampleStart
    println(1 + 2) // 3
    println(2_500_000_000L - 1L) // 2499999999
    println(3.14 * 2.71) // 8.5094
    println(10.0 / 3) // 3.3333333333333335
//sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

The result type depends on the types of the operands. Learn more in [](#mixed-numeric-expressions).

> You can override these operators in custom number classes.
> For more information, see [Operator overloading](operator-overloading.md).
>
{style="tip"}

### Integer division

Division between integer values always returns an integer result. The compiler discards the fractional part:

```kotlin
fun main() {
//sampleStart
    val intValue = 5 / 2
    println(intValue) // 2
    
    val longValue = 5L / 2
    println(longValue) // 2
//sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

To return a floating-point result, make at least one operand a `Float` or `Double`:

```kotlin
fun main() {
//sampleStart
    val a = 5 / 2.0
    println(a) // 2.5
    
    val b = 5 / 2.toDouble()
    println(b) // 2.5
//sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

> ### Data overflow

Reviewer: I just thought that data overflow can also occur with negation and conversion, not only arithmetic. Should this be its own subject lower on the page? In the "Type conversion" section we only show integer to floating point, but converting between integers also has a side effect.

Author: rewrote

| `Int`	   | 32          | -2,147,483,648 (-2<sup>31</sup>)             | 2,147,483,647 (2<sup>31</sup> - 1)             |
| `Long`	  | 64          | -9,223,372,036,854,775,808 (-2<sup>63</sup>) | 9,223,372,036,854,775,807 (2<sup>63</sup> - 1) |

**Declare integer values**

Kotlin supports the following literal forms for integer values:

* Decimals: `123`

> * `Long`, with the capital `L` suffix: `123L`

Reviewer 2: I guess this line should not be here. Suffixes are orthogonal to forms and mentioned later

Author: removed, thanks!

* Hexadecimals: `0x0F`
* Binaries: `0b00001011`

> Kotlin does not support octal literals.
>
{style="note"}

To declare `Int` and `Long` values, you can specify the type explicitly or let Kotlin infer it.
You can also use underscores to make numeric literals easier to read:

```kotlin
// Int
val one: Int = 1
val hexBytes = 0xFF_EC_DE_5E // Hexadecimal form
val bytes = 0b11010010_01101001_10010100_10010010 // Binary form

// Long
val oneBillion: Long = 1000000000 
val threeBillion = 3_000_000_000
```

To declare a `Long` value explicitly, you can also append the `L` suffix to the value:

```kotlin
val oneLong = 1L
```

To declare a `Byte` or `Short` value, specify the type explicitly:

```kotlin 
val oneByte: Byte = 1
val oneShort: Short = 1
```

> When you declare a numeric type explicitly, the compiler checks that the value
> fits in the range of that type.
>
> When you do not specify a numeric type, Kotlin infers `Int` if the 

> > value fits in the `Int` range. Otherwise, Kotlin infers `Long`. 

Reviewer 2: The structure of this part seem a bit off. Firstly it mentions "let Kotlin infer it", but explains how it is inferred only at the end. And "To declare X value, specify the type explicitly" is repeated separately for Long and Short/Byte.

Also it does not mention about inference for the following case:
```
fun foo(b: Byte)

foo(42) <- inferred to Byte, not Int
```

Would not it be better to make it as follows:
- To declare values of any numeric type, you can specify the type explicitly
- Specifically for Long, there is a suffix
- When you do not specify a numeric type, Kotlin trying to infer type from the context (example with `foo(b: Byte)`)
- If there is no information in the context, Kotlin infers `Int` if the value fits in the `Int` range. Otherwise, Kotlin infers `Long`.

The main idea is that it might be useful to mention this example of inference. On the other hand, it might be overcomplication for this chapter, so I would leave it on you to decide.

Author: Thanks, I agree with most of your points and will address them. However, yes, I don't believe the inference example is helpful here — it can overcomplicate things. So I would suggest focusing on the declaration part only and just mention inference as a note (what we have rn)

>
{style="note"}

If a value can be absent, use nullable types:

```kotlin
val maybeAbsent: Int? = null
```

## Floating-point types

For numbers with a fractional part, Kotlin provides `Float` and `Double`.

Floating-point types follow
the [IEEE 754 standard](https://en.wikipedia.org/wiki/IEEE_754).
`Float` reflects the _single precision_. `Double` reflects the _double precision_.

Floating-point types differ in size and precision:

| Type	    | Size (bits) | Significant bits | Exponent bits | Decimal digits |
|----------|-------------|------------------|---------------|----------------|
| `Float`	 | 32          | 24               | 8             | 6-7            |
| `Double` | 64          | 53               | 11            | 15-16          |    

**Declare floating-point values**

> To declare a floating-point literal, include a decimal point (`.`) or use an exponent notation. 

Reviewer 2: I guess it would be useful to demonstrate an exponent notation

By default, Kotlin infers floating-point literals as `Double`:

```kotlin
val pi = 3.14 // Double         
val oneDouble = 1.0 // Double   
```

To declare a `Float`, add the `f` or `F` suffix:

```kotlin
val eFloat = 2.7182818284f // Float   
// Actual value is 2.7182817

> ```

Reviewer 2: Probably it would be better with the following structure: 
- To declare a floating-point literal, include a decimal point (`.`) or use an exponent notation. 
- Example (with example for exponent notation)
- By default, Kotlin infers floating-point literals as `Double` + To declare a `Float`, add the `f` or `F` suffix
- Example

```kotlin
fun main() {
//sampleStart
    println(Double.MAX_VALUE * 2) // Infinity
//sampleEnd    
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

## Type conversion

Numeric types are not subtypes of one another. Kotlin requires explicit
conversions to avoid silent data loss and unexpected behavior.

For example, a function that expects `Double` cannot accept an `Int` or a `Float` value without conversion:

```kotlin
fun main() {
//sampleStart
    fun printDouble(x: Double) { 
        print(x) 
    }

    val x = 1.0
    val xInt = 1    
    val xFloat = 1.0f

    printDouble(x) // OK
    printDouble(xInt) // Error: argument type mismatch
    printDouble(xFloat) // Error: argument type mismatch
//sampleEnd
}
```
{kotlin-runnable="true" validate="false"}

All number types support conversions to other numeric types.
To convert a number to another type, use an explicit conversion function:

* `toByte()` (deprecated for [Float](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/-float/to-byte.html) and [Double](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/-double/to-byte.html))

> * `toShort()` (deprecated for [Float](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/-float/to-short.html) and [Double](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/-double/to-short.html))

Reviewer 2: Probably, it is an unnecessary overcomplication. They will be able to figure that out in that super rare scenario where they need it and the fix is described in the deprecation note.

Author: Makes sense, thanks! Removed.

* `toInt()`
* `toLong()`
* `toFloat()`
* `toDouble()`

For example, the following code converts an `Int` value to `Double`:

```kotlin
fun main() {
//sampleStart
    val intValue: Int = 1

>     val doubleValue= intValue.toDouble()

Reviewer 2: ```suggestion
    val doubleValue = intValue.toDouble()
```

    
    println(doubleValue) // 1.0
//sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

When you convert a floating-point value to an integer type, the compiler discards the fractional part:

```kotlin
fun main() {
//sampleStart
    val d: Double = 1.5
    val l: Long = d.toLong()
    
    println(l) // 1
//sampleEnd    
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

### Mixed numeric expressions

Kotlin does not support implicit conversion for assignments or function arguments. 
However, you can combine different numeric types in arithmetic expressions. In such cases, 

> Kotlin infers a result type based on the operand types, 

Reviewer 2: Probably it would be too detailed to describe rules, but could we provide a link to somewhere with it? (considering there is a reference to this section "The result type depends on the types of the operands. Learn more in [](#mixed-numeric-expressions).", it seems it should be either explained or referenced to somewhere else)

Also it is not actually an inference, so the following might be better: "Kotlin determines a result type ..."

Author: tbh, I don't see it being documented anywhere (or close enough to reference from this page).

Reviewer 2: Yes, probably it would be too detailed to explain it here, so it is fine

and arithmetic operators handle the conversion automatically:

```kotlin
val intNumber: Int = 1
val longNumber: Long = 1000
val result = intNumber + longNumber // 1001, Long
```

If you try to assign the result to a smaller type, the compiler reports an error:

```kotlin
val intNumber: Int = 1
val longNumber: Long = 1000
val result: Int = intNumber + longNumber 
// Error: Initializer type mismatch
```

> This behavior applies to expressions only. It does not replace explicit conversions
> when a specific target type is required.
>

> {style="note"}

Reviewer 2: I do not really understand what does this mean... Probably it should be something like: 
```
This behavior applies to arithmetic expressions only. As was already mentioned, Kotlin does not support implicit conversion for assignments or function arguments.
```

Author: rewrote

## Bitwise operations

Kotlin provides _bitwise operations_ for `Int` and `Long`. These operations are represented by functions that can
be called in [infix form](functions.md#infix-notation):

```kotlin
fun main() {
//sampleStart
    val x = 1
    
    println(x shl 2) // 4 
    println(x and 0x000FF000) // 0
//sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

Bitwise operations include:

> * `shl(bits)` – signed shift left

Reviewer 2: What is `bits`? It is not even the name of the argument in the declaration. Probably it is better to have just `shl` or use that actual name of the argument (`shl(bitCount)` for this case). The same for other operations.

Author: Thanks! I decided to drop the part in () completely to avoid confusion. I have a small question though: should we make a note that inv() is used like `a.inv()`, not as infix like other operations?

Reviewer 2: > should we make a note that inv() is used like a.inv(), not as infix like other operations

IMO, it is not necessary, as intuition about existence of it is enough. It is not that complicated to figure out the form, considering different signature.

On the other hand, what do you think about the following presentation:
- `a shl b` - ...
- `a shl b` - ...
- `a.inv()` - ...

It is simultaneously a demonstration of form, and avoids misleading `shl()` which might confuse reader about why there are no arguments

| `Int`	   | 32          | -2,147,483,648 (-2<sup>31</sup>)             | 2,147,483,647 (2<sup>31</sup> - 1)             |
| `Long`	  | 64          | -9,223,372,036,854,775,808 (-2<sup>63</sup>) | 9,223,372,036,854,775,807 (2<sup>63</sup> - 1) |

> **Declare integer values**

Reviewer: Not sure why this header is in bold instead of ###

Author: I didn't want it to be included in the TOC, but could not use <structure depth> since other titles should be displayed

Reviewer: Why not though? I think you have a good structure, but right now it's obscured — I very much enjoy that the boring table and actual declaration examples are clearly separate. Showing this in the ToC simply supports this.

More to the point, bold type has its own place — it's emphasis, on UI elements or parts of a sentence. Using it as a header muddies the users' general perception of our markup.

Author: Fair enough, fixed :)

Kotlin supports the following literal forms for integer values:

* Decimals: `123`
* Hexadecimals: `0x0F`
* Binaries: `0b00001011`

> Kotlin does not support octal literals.
>
{style="note"}

To declare a numeric value, specify the type explicitly: 

```kotlin
val one: Int = 1

> val oneBillion: Long = 1_000_000_000 // Use underscores to improve readability 

Reviewer: I'd add the examples of non-decimals with underscores here, they are even less apparent or intuitive.

Author: done

val oneByte: Byte = 1
val oneShort: Short = 1
```

You can also append the `L` suffix, to declare a `Long` value:

```kotlin
val oneLong = 1L
```

> When you declare a numeric type explicitly, the compiler checks that the value
> fits in the range of that type.
>
> When you do not specify a numeric type, Kotlin infers `Int` if the
> value fits in the `Int` range. Otherwise, Kotlin infers `Long`.
>

> {style="note"}

Reviewer: Is that a sidenote? To me this is a pretty straightforward explanation of what's happening when I declare stuff.

Author: I wanted to divide this theoretical information from the applicable one (how to declare). If I keep it as plain text, I feel like it could be lost and ignored due to the surrounding code blocks.

Reviewer: An admonition should be formatted as such if it's actually a break from the rest of the text. Here "...specify the type explicitly: <...> When you declare a numeric type explicitly" is a clear sequence for me, that's why it looks strange that it's visually broken up. Admonition is metainformation; using it for visual effect tends to lead to warning fatigue, when people just start skipping everything that's not the main body.

To help it not get lost we can add a tailored example here. If it would get lost, it's because we have a clear "theory → immediate example" pattern on the page, and here it's not working.

And, last but not least, I wouldn't say it's purely theoretical :) one needs to know how inference works to handle it properly, so to me this is straightforwardly useful.

Author: Fixed, added examples as well.

If a value can be absent, use nullable types:

```kotlin
val maybeAbsent: Int? = null
```

## Floating-point types

For numbers with a fractional part, Kotlin provides `Float` and `Double`.

Floating-point types follow
the [IEEE 754 standard](https://en.wikipedia.org/wiki/IEEE_754).
`Float` reflects the _single precision_. `Double` reflects the _double precision_.

Floating-point types differ in size and precision:

| Type	    | Size (bits) | Significant bits | Exponent bits | Decimal digits |
|----------|-------------|------------------|---------------|----------------|
| `Float`	 | 32          | 24               | 8             | 6-7            |
| `Double` | 64          | 53               | 11            | 15-16          |    

> **Declare floating-point values**

Reviewer: Again, should be a proper heading IMO.

> To declare a floating-point literal, include a decimal point (`.`) or use an exponent notation:

Reviewer: ```suggestion
To declare a floating-point literal, include a decimal point (`.`) or use the exponential notation:
```

Author: I think **exponential notation** sounds less idiomatic? But now I also feel like the article should be omitted here

```kotlin
val pi = 3.14

> val avogadro = 6.02214076e23  

Reviewer: ```suggestion
val avogadro = 6.02214076e23
```

```

By default, Kotlin infers floating-point literals as `Double`. 
To declare a `Float`, add the `f` or `F` suffix:

```kotlin
val pi = 3.14 // Double         
val eFloat = 2.7182818284f // Float
```

> Kotlin rounds a `Float` literal that contains more precision than `Float` can store.
>
{style="note"}

> If a value can be absent, use nullable types:

Reviewer: I don't know whether this is useful for numbers specifically. I think this is one of the things Kotlin beginners bump against on every step, so talking about this explicitly may not be too useful.

Author: I would like to keep it for consistency with other Types pages, tbh

Reviewer: I would like to remove it from all Types pages :) but I don't have a strong argument, let them bump some more.

{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

All number types support conversions to other types:
Here, the result wraps around because the value no longer fits in `Int`.

* `toByte(): Byte` (deprecated for [Float](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/-float/to-byte.html) and [Double](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/-double/to-byte.html))
* `toShort(): Short`
* `toInt(): Int`
* `toLong(): Long`
* `toFloat(): Float`
* `toDouble(): Double`
> The compiler does not automatically produce an error when integer overflow occurs.
>
{style="note"}

In many cases, there is no need for explicit conversion because the type is inferred from the context,
and arithmetical operators are overloaded to handle conversions automatically. For example:
However, since floating-point types follow the
[IEEE 754 Standard](https://en.wikipedia.org/wiki/IEEE_754), very large results can become `Infinity`:

```kotlin
fun main() {
//sampleStart
    val l = 1L + 3       // Long + Int => Long
    println(l is Long)   // true
//sampleEnd
    println(Double.MAX_VALUE * 2) // Infinity
//sampleEnd    
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.5"}

> ### Reasoning against implicit conversions

Reviewer: I would argue that this section, while more technical, is small enough to be left inside the "Type conversion" section for those wondering why implicit conversions are not supported. It's not like we have a more in-depth explanation of this elsewhere — probably because here's where it usually matters.

Author: I tired to focus on why we prefer explicit conversion over implicit one. A "glass half full" kinda logic :D 
And find the example in this section to be tricky for novice users. tbh

Reviewer: Why explicit is good / why implicit is bad is kind of sides of the same coin, but I think we should describe both sides to show the coin as a whole. We can even hide it under the cut to not push it on people who just want to get through the basics.

My point is also that this page is not solely for novice users — experienced developers who are looking for "why numbers in Kotlin work the way they work" also deserve some attention.

That said, this conversation may be indeed too academic to remain in the docs at all, so it's your call.

Author: Yeah, I feel like it can be turned into a good blog post we can reference from here

>     printDouble(xFloat) // Error: argument type mismatch

Reviewer: Maybe also tuck in the `val one: Double = 1` example here — it's also strictly speaking a type mismatch, but may be not apparent.

Author: I don't want to make this example too big. WDYT about replacing float with a double?

Reviewer: The point is in the fact that cross-type assignment _also_ fails, which is not obvious; the fact that the function call fails we explicitly mention before the sample. I wouldn't say that a single line is too big of a price for that.

Author: added

> All number types support conversions to other numeric types.

Reviewer: The question I immediately have is whether number types are the same as numeric types — when referring to the same concept, I'd use the same term.

Author: Good catch, thanks!

> * `toDouble()`

Reviewer: I know that the list has been here already, but I can't decide whether I like it. It takes a lot of space on the page and communicates basically the same thing as the sentence before it. Maybe we should put these into parentheses at the end of the sentence?

Author: Agree that it takes too much space, but I feel like it acts as a divider between why and how, making it easier to find the how part if you don't care/already know about the why.

> {style="note"}

Reviewer: I am not sure what this note is about, what else this behavior can apply to. Maybe this part would make more sense — and not require a note —inside the "Arithmetic operations﻿" section? Since the whole "Mixed numeric expressions" part seems to be about arithmetic expressions.

Author: Yes, I feel like that note can be confusing. Thanks!

> be called in [infix form](functions.md#infix-notation):

Reviewer: I would suggest simply going "...by a set of [infix functions]():" because "infix form" is not actually defined in functions.md.

Author: Rewrote, but clearly stated `inv()` since it is not an infix function.

* `NaN` is considered equal to itself
* `NaN` is considered greater than any other element including `POSITIVE_INFINITY`
* `-0.0` is considered less than `0.0`

The following example shows the difference between operands statically typed as floating-point numbers
and operands used through generic types:

```kotlin
//sampleStart  
fun generalizedEquals(a: Any, b: Any): Boolean {
  return a == b
}

fun main() {
  // Operands statically typed as floating-point numbers
  println(Double.NaN == Double.NaN) // false
  println(0.0 == -0.0) // true

  // Operands used through a non-floating-point static type
  println(generalizedEquals(Double.NaN, Double.NaN)) // true
  println(generalizedEquals(0.0, -0.0)) // false
//sampleEnd  
}
```

> {kotlin-runnable="true" kotlin-min-compiler-version="1.3" id="kotlin-numbers-floating-comp"}

Reviewer: The sample uses two-space indent instead of four.

Author: good catch, thanks:)

represented as an object.

The JVM applies a [memory optimization technique](https://docs.oracle.com/javase/specs/jls/se22/html/jls-5.html#jls-5.1.7)
to small numbers by caching their boxed representations. As a result,
boxed numbers with the same value can be [referentially equal](equality.md#referential-equality).

For example, the JVM caches boxed `Integer` values in the range `-128` to `127`. Therefore, the following
code returns `true`:

```kotlin
fun main() {
//sampleStart
    val score: Int = 100
    val savedScore: Int? = score
    val displayedScore: Int? = score
    
    println(savedScore === displayedScore) // true
//sampleEnd
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}

For values outside the cached range, boxed values are separate objects. In that case,
they are not referentially equal, even if their values are [structurally equal](equality.md#structural-equality).
For this reason, use `==` to compare numeric values:

```kotlin
fun main() {
//sampleStart
    val score: Int = 10000
    val savedScore: Int? = score
    val displayedScore: Int? = score

    println(savedScore == displayedScore) // true

>     println(savedScore === displayedScore) // false

Reviewer: That's nitpicky, but I'd put the bad example before the good one. It emulates making a mistake: the fact that === should not be used is a "surprise", and the correct solution is then revealed. When we show the correct one first and then show a mistake, the mistake doesn't feel relevant.

Author: That's a good point! Thanks

* `xor(bits)` – bitwise **XOR**
* `inv()` – bitwise inversion

## Floating-point number comparison

In Kotlin, floating-point comparison depends on the static type of the operands.

> When the operands are statically known to be `Float` or `Double` (including their nullable counterparts),

Reviewer 2: "nullable counterparts"? AFAIU, it is not possible to make arithmetic over nullable versions. So probably it is better to just remove the text in parenthesis

Author: thanks, fixed.

operations on the numbers and the range that they form
follow the [IEEE 754 Standard for Floating-Point Arithmetic](https://en.wikipedia.org/wiki/IEEE_754).

However, in generic use cases (such as `Any`, `Comparable<...>`, or `Collection<T>`), behavior differs for
operands that are not statically typed as floating-point numbers. In these cases, Kotlin
uses the `equals()` and `compareTo()` implementations for `Float` and `Double`. 

As a result:

* `NaN` is considered equal to itself
* `NaN` is considered greater than any other element including `POSITIVE_INFINITY`
* `-0.0` is considered less than `0.0`

The following example shows the difference between operands statically typed as floating-point numbers
and operands used through generic types:

```kotlin
fun main() {
    //sampleStart
    // Operand statically typed as floating-point number
    println(Double.NaN == Double.NaN) // false
    
    // Operand NOT statically typed as floating-point number
    println(listOf(Double.NaN) == listOf(Double.NaN)) // true

    // Operand statically typed as floating-point number
    println(0.0 == -0.0) // true
    
    // Operand NOT statically typed as floating-point number

>     println(listOf(0.0) == listOf(-0.0)) // false

Reviewer 2: Not sure if this example is easy to understand. It assumes that reader already understand how generics in Kotlin works (about erasure). Would not it better to demonstrate it with the following:
```
fun generalizedEquals(a: Any, b: Any) {
    return a == b
}

fun main() {
    ...
    println(generalizedEquals(Double.NaN, Double.NaN))
    ...
}
```

Author: makes sense, updated.
